mypyc/genops.py

```diff
@@ -991,39 +1054,64 @@ def visit_mypy_file(self, mypyfile: MypyFile) -> None:
+                # If the class is a non-extension class, the undecorated method is the
+                # generated callable class
+                assert func_reg is not None
+                orig_func = func_reg
+            else:
+                # Read the PyTypeObject representing the class, get the callable object
+                # representing the non-decorated method
+                typ = self.load_native_type_object(cdef.fullname)
+                orig_func = self.py_get_attr(typ, helper_name, fdef.line)
+
+            # Decorate the non-decorated method
             decorated_func = self.load_decorated_func(fdef, orig_func)
 
-            # Set the callable object representing the decorated method as an attribute of the
-            # class.
-            class_ir.attributes[name] = decorated_func.type
-            self.primitive_op(py_setattr_op,
-                              [typ, self.load_static_unicode(name), decorated_func], fdef.line)
+            if class_ir.is_non_ext:
+                func_reg = decorated_func
+
+            else:
+                # Set the callable object representing the decorated method as an attribute of the
+                # extension class.
+                class_ir.attributes[name] = decorated_func.type
+                self.primitive_op(py_setattr_op,
+                                  [typ, self.load_static_unicode(name), decorated_func], fdef.line)
 
         if fdef.is_property:
-            # If there is a property setter, it will be processed after the getter,
-            # We populate the optional setter field with none for now.
-            assert name not in class_ir.properties
-            class_ir.properties[name] = (func_ir, None)
+            if class_ir.is_non_ext:
+                # Wrap the method in a call to property if it is in a non-extension class
+                prop = self.load_module_attr_by_fullname('builtins.property', fdef.line)
+                assert func_reg is not None
+                func_reg = self.py_call(prop, [func_reg], fdef.line)
+            else:
+                # If there is a property setter, it will be processed after the getter,
+                # We populate the optional setter field with none for now.
+                assert name not in class_ir.properties
+                class_ir.properties[name] = (func_ir, None)
 
-        elif fdef in self.prop_setters:
+        # TODO: Support property setters in non-extension classes
+        elif fdef in self.prop_setters and not class_ir.is_non_ext:
             # The respective property getter must have been processed already
             assert name in class_ir.properties
             getter_ir, _ = class_ir.properties[name]
             class_ir.properties[name] = (getter_ir, func_ir)
 
-        if func_ir.decl.is_prop_setter:
+        # If the classs is a non-extension class, add the function to the classes dict.
+        if class_ir.is_non_ext:
+            assert func_reg is not None
```

Comment:
TBH I don't really like this kind of "non-local" control flow. up ti this point not much code is shared between extensions and non-extension cases. I would just have to methods for these two cases each having more linear flow of logic.